src/Build/Construction/Solution/SolutionFile.cs

```diff
@@ -1211,6 +1218,50 @@ internal void ParseNestedProjects()
 
                 proj.ParentProjectGuid = parentProjectGuid;
             } while (true);
+
+            ProjectFileErrorUtilities.VerifyThrowInvalidProjectFile(ContainsNoCyclicalNesting(), "SubCategoryForSolutionParsingErrors",
+                new BuildEventFileInfo(FullPath, _currentLineNumber, 0), "SolutionContainsCyclicallyNestedProjectFiles", _solutionFile);
+        }
+
+        private bool ContainsNoCyclicalNesting()
+        {
+            SearchState[] parents = new SearchState[_projectsInOrder.Count];
+            for (int a = 0; a < _projectsInOrder.Count; a++) {
+                parents[a] = SearchState.Unchecked;
+            }
```

Comment:
Change the enum so that `Unchecked` has the backing value `0` (is listed first, optionally with `= 0`). Then the default-initialized array will do this for you.

[sharplab demonstration of this](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUBgBjwwGYACAbzzJrIAcwBLANwENgBTMjqCAWzIBlDqzABjABaDg7LiDIAjAJ6dqtKrlpayAVSiSOYgNYc4ZALxkCMNdrIBBfRwDOwAPZhnAYTdQZjKAA5Ny8lMQAbFxtNOzIvCUMTOFsyAF88FIx0ShStYVFJaVkAbQBdelEeYGcLMigOAHchEXEpGU5izAJSgG5c2gAzDzIACgDgMkZagh7JsgAeCrAq5wA6ABkeAHNgCVnGREQASn6aDVjtVEwAThG6Sr9nYsZSo76Yu3SPtLxUoA)